### PR TITLE
Action before and after subscribers

### PR DESCRIPTION
**Pull request by wa3l**
See issue https://github.com/vuejs/vuex/issues/1098. Thanks to @ktsn for proposing the api.

Add the ability to specify before and after action subscribers that behave as follows: 

- `before`: is called before the action like subscribers behave now. If you only pass a function to `subscribeAction` then it's assumed that it's the before subscriber for backwards compatibility.
- `after`: is called after the action _resolves_. 

Examples:

```js
// calls foo() before every action
store.subscribeAction(foo)

// calls foo() before every action
store.subscribeAction({
  before: foo
})

// calls foo() after an action resolves
store.subscribeAction({
  after: foo
})

// calls foo() before every action and bar() after the action resolves
store.subscribeAction({
  before: foo,
  after: bar,
})
```

Let me know if you have any suggestions/improvements!

### CODE REVIEW

File: test/unit/modules.spec.js [hunk 1]
```diff
@@ -668,6 +668,37 @@ describe('Modules', () => {
         store.state
       )
     })
+
+    it('action before/after subscribers', (done) => {
+      const beforeSpy = jasmine.createSpy()
+      const afterSpy = jasmine.createSpy()
+      const store = new Vuex.Store({
+        actions: {
+          [TEST]: () => new Promise(resolve => resolve())
```
Comment (ktsn) [inline test/unit/modules.spec.js]: This can be simplified with `Promise.resolve()` :slightly_smiling_face:

**Comment by ktsn** [comment]
Also, can we update TypeScript type declaration?
https://github.com/vuejs/vuex/blob/dev/types/index.d.ts#L20

If you are not sure how to update it, I'll do it later.

File: test/unit/modules.spec.js (hunk 1 above)
Comment (wa3l) [reply test/unit/modules.spec.js]: oh duh. Will update!

**Comment by wa3l** [comment]
@ktsn I have no experience with typescript, but the syntax seemed straightforward enough. Let me know if you have suggestions!

**Comment by ktsn** [comment]
@wa3l Well done! 👍

**Comment by wa3l** [comment]
@ktsn glad to hear it! Next steps?

**Comment by KindWizzard** [comment]
Hi guys! How are things going with this improvement? The last comment was added 25 days ago :) Looks like it's freezed. Will it be released?

**Comment by ktsn** [comment]
Waiting for @yyx990803's review.

**Comment by wa3l** [comment]
@yyx990803 any updates on this?

**Comment by KindWizzard** [comment]
Guys, please, if you don't want to approve this request, can you suggest some other solution? Me and my team still using fork. But, you know, it's not looks good, cause Vuex is part of Vue, Vue is part of Nuxt :)

**Comment by Vincent1993** [comment]
@yyx990803 please, It is very useful to me, waiting for you review

**Comment by wa3l** [comment]
@ktsn is there no way around Evan having to review every single PR? Seems like there is enough appetite for this PR but I don't see it getting merged if one person has to review every pull request on several different libraries...

**Comment by chearon** [comment]
This would be really useful for integration tests too. Right now I wish I could have something like

```javascript
closeDialog();
waitForStoreActions();
expect(...);
```

The `waitForStoreActions()` part could be implemented by using this PR.

FWIW, I would have just expected `subscribeAction` to be changed to have a promise as the third argument:

```javascript
store.subscribeAction(async (action, state, done) => {
  // this is already like the before() part of the PR
  await done;
  // now this is like the after() part of this PR
});
```

It might be a breaking change though since it has to be called _after_ all the actions are fired.

**Comment by marlonbarcarol** [comment]
Looks like that it's done to be merged :)

**Comment by wa3l** [comment]
@yyx990803 any timeline on merging this?

**Comment by leoyli** [comment]
Can't wait for this!

**Comment by leoyli** [comment]
Hi @yyx990803, @ktsn, I'm just wonder if anything have to be done in order to pass this through.  Or is this feature will have some other concerns (e.g. conflicts with the future movement).

I can see the last commit is done in very early this year and have been fully reviewed in June, but still it is not merged and sitting there.  Please, please, please pointing some directions, concerns, or whatever so the community become more healthy and moving.  I definitely don't know why or if something haven't been done. May be the documentation or using example?  At least the big heads have to tell us so we can contribute.  Thanks!

**Comment by wa3l** [comment]
@leoyli I'm equally frustrated. I don't know if what happened with this PR is typical of other ones, but it certainly discouraged me from contributing any further. I've used Vue (and vuex, vue-router, vue-meta) everyday for the past couple of years, so I'm happy to pay some of that karma back to the community, but I just don't see how that's possible with the current process.

**Comment by gitlab-winnie** [comment]
@ktsn Based on your comment (https://github.com/vuejs/vuex/pull/1115#issuecomment-361490858) this awaited review by Evan who approved this in June (https://github.com/vuejs/vuex/pull/1115#pullrequestreview-125994968). Could you please comment on the current state of this pull request?

**Comment by Sunhat** [comment]
Seriously over a year since this was originally proposed?

**Comment by yyx990803** [comment]
Sorry for letting this sit, will do a release next week.

**Comment by Sunhat** [comment]
Thank you so much for the quick response! Excited for my loading screens!

**Comment by wa3l** [comment]
@yyx990803 any updates on the release?

**Comment by yyx990803** [comment]
Out in https://github.com/vuejs/vuex/releases/tag/v3.1.0

**Pull request merged**